doc/15-troubleshooting.md

```diff
@@ -230,6 +232,91 @@ icinga2 7764 nagios  mem    REG                8,5   744776 2631636 /usr/lib/x86
 #
 ```
 
+### Optimise apply rules and group assign conditions
+
+#### Remove actually unused apply rules
+
+If `icinga2 daemon -C` warns you like shown below and the respective apply rule
+should indeed create no objects, consider removing it or commenting it out.
+
+```
+[...] warning/ApplyRule: Apply rule '...' (...) for type '...' does not match anywhere!
+```
+
+At least consider changing the `assign where` filter from, say
+`assign where my_complex_filter()`, to `assign where false && my_complex_filter()`.
+`&&` will evaluate `false` first and in an instant.
+If it's not true, `&&` will "return" `false` and skip `my_complex_filter()`.
+
+Same for `assign where` conditions in groups. In this case removing just
+the `assign where` line(s) is enough if you'd like to keep the group itself.
+
+#### Avoid creating single objects via apply rules
+
+If possible, replace constructs like the immediately following with the below one.
+
+```
+apply Service "syspatch" {
+	// ...
+	assign where host.name == "firewall"
+}
+```
+
+```
+object Service "syspatch" {
+	// ...
+	host_name = "firewall"
+}
+```
+
+This way Icinga won't have to evaluate `host.name == "firewall"`
+for all of your hosts.
+
+At least upgrade to Icinga v2.13.6+ which optimizes apply rules as shown above.
+(In this case such `assign where` filters even save time compared to e.g.
+`assign where host.vars.os == "OpenBSD"` if such hosts are a minority.
+Useful if the above recommendation is not an option.)
```

Comment:
It is unclear what this is trying to say. `assign where host.name == "firewall"` and `assign where host.vars.os == "OpenBSD"` do different things and the first isn't faster than `object` + `host_name = "firewall"` so why recommend using it right after recommending the opposite?

Maybe you're trying something like the following?

> If an apply rule generates a set of objects that is constant or changes infrequently so that it is feasible to create the objects individually, do so as it avoid evaluating the `assign where` expression many times.